framework/include/materials/MaterialPropertyRegistry.h

```diff
@@ -0,0 +1,81 @@
+//* This file is part of the MOOSE framework
+//* https://www.mooseframework.org
+//*
+//* All rights reserved, see COPYRIGHT for full restrictions
+//* https://github.com/idaholab/moose/blob/master/COPYRIGHT
+//*
+//* Licensed under LGPL 2.1, please see LICENSE for details
+//* https://www.gnu.org/licenses/lgpl-2.1.html
+
+#pragma once
+
+#include <string>
+#include <unordered_map>
+#include <vector>
+
+class MaterialData;
+class MaterialPropertyInterface;
+class MaterialPropertyStorage;
+
+/**
+ * Registry class for material property IDs and names.
+ *
+ * Not thread safe.
+ */
+class MaterialPropertyRegistry
+{
+public:
+  /**
+   * @return True if a material property is registered with the name \p name
+   */
+  bool hasProperty(const std::string & name) const { return _name_to_id.count(name); }
+
+  /**
+   * @return True if a material property is registered with the ID \p id
+   */
+  bool hasProperty(const unsigned int id) const { return id < _id_to_name.size(); }
+
+  /**
+   * Key that restricts writing data to the registry.
+   */
+  class WriteKey
+  {
+    friend class MaterialPropertyStorage;
+    WriteKey() {}
+    WriteKey(const WriteKey &) {}
+  };
+
+  /**
+   * @return The property ID for the given name, adding the property and
+   * creating a new ID if it hasn't already been created.
+   *
+   * Protected with the MaterialPropertyRegistry::WriteKey.
+   */
+  unsigned int addOrGetID(const std::string & name, const WriteKey);
+
+  /**
+   * @return The property ID for the property with the name \p name
+   *
+   * Will not create an ID if one does not exist, unlike addOrGetPropertyId
+   */
+  unsigned int getID(const std::string & name) const;
+
+  /**
+   * @return The property name for the property with the ID \p id
+   */
+  const std::string & getName(const unsigned int id) const;
+
+  /**
+   * @return The mapping of material property name to material property ID
+   */
+  const std::unordered_map<std::string, unsigned int> & getNamesToIDs() const
```

Comment:
Do we use this anywhere yet?  "Just get a reference to the container" is the sort of API that I tend to make without thinking and then regret down the road.

Reply:
Yep. In ComputeFVFluxThread

Reply:
I do prefer at least "return an iterator range of a type we control" over "just return a reference to the container" (imagine how stuck `DistributedMesh` would have been otherwise...) but if it's too much hassle this is fine as is.